File: RealtimeTranslator/Realtime/ReconnectBudget.swift
import Foundation

/// 再接続の待ち時間と打ち切り条件。shared/fixtures/v1/reconnect.json の policy と同じ値を持つ。
struct ReconnectPolicy: Sendable, Equatable {
    var initialBackoff: Duration
    var backoffMultiplier: Int
    var maxBackoff: Duration
    var jitterMax: Duration
    var maxAttempts: Int
    var totalBudget: Duration
    var stablePeriod: Duration

    static let `default` = ReconnectPolicy(
        initialBackoff: .milliseconds(500),
        backoffMultiplier: 2,
        maxBackoff: .seconds(8),
        jitterMax: .milliseconds(250),
        maxAttempts: 5,
        totalBudget: .seconds(120),
        stablePeriod: .seconds(30)
    )

    /// attempt 回目（1 始まり）の指数バックオフ。maxBackoff で頭打ち。
    func backoff(forAttempt attempt: Int) -> Duration {
        precondition(attempt >= 1, "attempt must be 1-based")
        var value = initialBackoff
        for _ in 1..<attempt {
            value *= backoffMultiplier
            if value >= maxBackoff {
                return maxBackoff
            }
        }
        return min(value, maxBackoff)
    }
}

enum ReconnectDecisionKind: Sendable, Equatable {
    case wait
    case attemptLimit
    case budgetExhausted
}

struct ReconnectDecision: Sendable, Equatable {
    let kind: ReconnectDecisionKind
    let attempt: Int
    let backoff: Duration
    let jitter: Duration

    var delay: Duration { backoff + jitter }
}

/// 連続障害の経過を単調クロックで追い、試行回数と総予算の両方で再接続を打ち切る。
///
/// - Listening に入っただけでは試行回数をリセットしない。
/// - Listening が `stablePeriod` 以上続いた後の失敗で、試行回数と予算を新しく始める。
/// - 総予算は試行回数と独立に判定し、先に尽きた方で停止する。
struct ReconnectBudget: Sendable {
    typealias Now = @Sendable () -> Duration
    typealias Jitter = @Sendable (Duration) -> Duration

    /// 単調クロック（ContinuousClock）。壁時計の補正やスリープ復帰に影響されない。
    static let continuousNow: Now = {
        let origin = ContinuousClock.now
        return { origin.duration(to: .now) }
    }()

    static let randomJitter: Jitter = { max in
        let maxNanoseconds = Self.nanoseconds(max)
        guard maxNanoseconds > 0 else { return .zero }
        return .nanoseconds(Int64.random(in: 0...maxNanoseconds))
    }

    let policy: ReconnectPolicy
    private let now: Now
    private let jitter: Jitter

    private var attempt = 0
    private var outageStart: Duration?
    private var listeningSince: Duration?

    init(
        policy: ReconnectPolicy = .default,
        now: @escaping Now = ReconnectBudget.continuousNow,
        jitter: @escaping Jitter = ReconnectBudget.randomJitter
    ) {
        self.policy = policy
        self.now = now
        self.jitter = jitter
    }

    /// 現在の試行回数（バナー表示用）。
    var currentAttempt: Int { attempt }

    /// 新しい録音世代の開始で呼ぶ。過去の障害履歴を捨てる。
    mutating func reset() {
        attempt = 0
        outageStart = nil
        listeningSince = nil
    }

    /// Listening に入った時刻を記録する。これだけでは試行回数をリセットしない。
    mutating func recordListening() {
        listeningSince = now()
    }

    /// 失敗を記録し、待つか打ち切るかを返す。
    mutating func recordFailure() -> ReconnectDecision {
        let current = now()

        if let listeningSince {
            self.listeningSince = nil
            if current - listeningSince >= policy.stablePeriod {
                attempt = 0
                outageStart = nil
            }
        }

        let start = outageStart ?? current
        outageStart = start
        if current - start >= policy.totalBudget {
            return ReconnectDecision(kind: .budgetExhausted, attempt: attempt, backoff: .zero, jitter: .zero)
        }

        if attempt >= policy.maxAttempts {
            return ReconnectDecision(kind: .attemptLimit, attempt: attempt, backoff: .zero, jitter: .zero)
        }

        attempt += 1
        let backoff = policy.backoff(forAttempt: attempt)
        var jitterValue = jitter(policy.jitterMax)
        if jitterValue < .zero || jitterValue > policy.jitterMax {
            jitterValue = .zero
        }
        return ReconnectDecision(kind: .wait, attempt: attempt, backoff: backoff, jitter: jitterValue)
    }

    static func nanoseconds(_ duration: Duration) -> Int64 {
        let components = duration.components
        return components.seconds * 1_000_000_000 + components.attoseconds / 1_000_000_000
    }
}

File: windows/src/RealtimeTranslator.Core/Realtime/ReconnectBudget.cs
using System;

namespace RealtimeTranslator.Core.Realtime;

/// <summary>再接続の backoff・回数・総予算・安定期間。<c>shared/fixtures/v1/reconnect.json</c> が正本。</summary>
public sealed record ReconnectPolicy(
    TimeSpan InitialBackoff,
    int BackoffMultiplier,
    TimeSpan MaxBackoff,
    TimeSpan JitterMax,
    int MaxAttempts,
    TimeSpan TotalBudget,
    TimeSpan StablePeriod)
{
    public static ReconnectPolicy Default { get; } = new(
        InitialBackoff: TimeSpan.FromMilliseconds(500),
        BackoffMultiplier: 2,
        MaxBackoff: TimeSpan.FromSeconds(8),
        JitterMax: TimeSpan.FromMilliseconds(250),
        MaxAttempts: 5,
        TotalBudget: TimeSpan.FromSeconds(120),
        StablePeriod: TimeSpan.FromSeconds(30));

    /// <summary>jitter を含まない attempt 回目（1 始まり）の backoff。</summary>
    public TimeSpan BackoffFor(int attempt)
    {
        ArgumentOutOfRangeException.ThrowIfLessThan(attempt, 1);
        var factor = Math.Pow(BackoffMultiplier, attempt - 1);
        var scaled = InitialBackoff.Ticks * factor;
        if (scaled >= MaxBackoff.Ticks)
        {
            return MaxBackoff;
        }

        return TimeSpan.FromTicks((long)scaled);
    }
}

public enum ReconnectDecisionKind
{
    /// <summary>backoff だけ待ってから再接続する。</summary>
    Wait,
    /// <summary>attempt が上限を超えた。<c>error.reconnectLimit</c>。</summary>
    AttemptLimit,
    /// <summary>連続障害の開始から総予算を使い切った。<c>error.reconnectBudgetExhausted</c>。</summary>
    BudgetExhausted,
}

public readonly record struct ReconnectDecision(
    ReconnectDecisionKind Kind,
    int Attempt,
    TimeSpan Backoff,
    TimeSpan Jitter)
{
    public TimeSpan Delay => Backoff + Jitter;
}

/// <summary>
/// 連続障害の経過を単調クロック（<see cref="TimeProvider.GetTimestamp"/>）で追う。
/// Listening に入っただけではリセットせず、<see cref="ReconnectPolicy.StablePeriod"/> 以上維持したあとの失敗だけが
/// attempt と予算を作り直す。壁時計の変化は影響しない。スレッド安全ではないので呼び出し側で直列化する。
/// </summary>
public sealed class ReconnectBudget
{
    private readonly ReconnectPolicy _policy;
    private readonly TimeProvider _timeProvider;
    private readonly Func<TimeSpan, TimeSpan> _jitter;

    private int _attempt;
    private long? _outageStartTimestamp;
    private long? _listeningSinceTimestamp;

    public ReconnectBudget(
        ReconnectPolicy? policy = null,
        TimeProvider? timeProvider = null,
        Func<TimeSpan, TimeSpan>? jitter = null)
    {
        _policy = policy ?? ReconnectPolicy.Default;
        _timeProvider = timeProvider ?? TimeProvider.System;
        _jitter = jitter ?? DefaultJitter;
    }

    public ReconnectPolicy Policy => _policy;

    /// <summary>現在の連続障害内で消費した attempt。</summary>
    public int Attempt => _attempt;

    /// <summary>録音開始時。すべて忘れる。</summary>
    public void Reset()
    {
        _attempt = 0;
        _outageStartTimestamp = null;
        _listeningSinceTimestamp = null;
    }

    /// <summary>Listening へ入った。安定期間の起点だけを記録し、attempt / 予算は触らない。</summary>
    public void RecordListening()
    {
        _listeningSinceTimestamp = _timeProvider.GetTimestamp();
    }

    /// <summary>接続が失われた。次に待つ時間、または停止理由を返す。</summary>
    public ReconnectDecision RecordFailure()
    {
        var now = _timeProvider.GetTimestamp();

        if (_listeningSinceTimestamp is { } listeningSince)
        {
            var listened = _timeProvider.GetElapsedTime(listeningSince, now);
            _listeningSinceTimestamp = null;
            if (listened >= _policy.StablePeriod)
            {
                _attempt = 0;
                _outageStartTimestamp = null;
            }
        }

        _outageStartTimestamp ??= now;
        var elapsed = _timeProvider.GetElapsedTime(_outageStartTimestamp.Value, now);
        if (elapsed >= _policy.TotalBudget)
        {
            return new ReconnectDecision(ReconnectDecisionKind.BudgetExhausted, _attempt, TimeSpan.Zero, TimeSpan.Zero);
        }

        if (_attempt >= _policy.MaxAttempts)
        {
            return new ReconnectDecision(ReconnectDecisionKind.AttemptLimit, _attempt, TimeSpan.Zero, TimeSpan.Zero);
        }

        _attempt += 1;
        var backoff = _policy.BackoffFor(_attempt);
        var jitter = _jitter(_policy.JitterMax);
        if (jitter < TimeSpan.Zero || jitter > _policy.JitterMax)
        {
            jitter = TimeSpan.Zero;
        }

        return new ReconnectDecision(ReconnectDecisionKind.Wait, _attempt, backoff, jitter);
    }

    private static TimeSpan DefaultJitter(TimeSpan max) =>
        max <= TimeSpan.Zero
            ? TimeSpan.Zero
            : TimeSpan.FromTicks(Random.Shared.NextInt64(0, max.Ticks + 1));
}
